File: internal/sandbox-sidecar/filesystem/filesystem_test.go
// Copyright The Isola Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package filesystem

import (
	"bytes"
	"crypto/rand"
	"errors"
	"log/slog"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"sync"
	"syscall"
	"time"

	"github.com/danielgtaylor/huma/v2"
	"github.com/danielgtaylor/huma/v2/humatest"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"

	sandboxsidecar "github.com/isola-run/isola/internal/sandbox-sidecar"
	"github.com/isola-run/isola/internal/sandbox-sidecar/proc"
)

var _ = Describe("Filesystem", func() {
	Describe("GET /filesystem", func() {
		It("reads file with absolute path", func() {
			content := []byte("hello world")
			hostPath := filepath.Join(testRootDir, "/tmp/readable.txt")
			Expect(os.MkdirAll(filepath.Dir(hostPath), 0750)).To(Succeed())
			Expect(os.WriteFile(hostPath, content, 0600)).To(Succeed())

			resp := doGet("/v1/filesystem?path=/tmp/readable.txt")

			Expect(resp.Code).To(Equal(http.StatusOK))
			Expect(resp.Body.Bytes()).To(Equal(content))
		})

		It("reads file with relative path", func() {
			content := []byte("relative file content")
			hostPath := filepath.Join(testRootDir, testCwd, "myreadfile.txt")
			Expect(os.WriteFile(hostPath, content, 0600)).To(Succeed())

			resp := doGet("/v1/filesystem?path=myreadfile.txt")

			Expect(resp.Code).To(Equal(http.StatusOK))
			Expect(resp.Body.Bytes()).To(Equal(content))
		})

		It("reads binary content correctly", func() {
			content := []byte{0x00, 0xFF, 0x80, 0x01, 0xFE, 0x7F}
			hostPath := filepath.Join(testRootDir, "/tmp/binary.dat")
			Expect(os.MkdirAll(filepath.Dir(hostPath), 0750)).To(Succeed())
			Expect(os.WriteFile(hostPath, content, 0600)).To(Succeed())

			resp := doGet("/v1/filesystem?path=/tmp/binary.dat")

			Expect(resp.Code).To(Equal(http.StatusOK))
			Expect(resp.Body.Bytes()).To(Equal(content))
		})

		It("returns 404 for nonexistent file", func() {
			resp := doGet("/v1/filesystem?path=/tmp/nonexistent.txt")

			Expect(resp.Code).To(Equal(http.StatusNotFound))
		})

		It("returns 400 for directory path", func() {
			resp := doGet("/v1/filesystem?path=/workspace")

			Expect(resp.Code).To(Equal(http.StatusBadRequest))
		})

		It("returns 400 for FIFO (named pipe)", func() {
			fifoPath := filepath.Join(testRootDir, "/tmp/test.fifo")
			Expect(os.MkdirAll(filepath.Dir(fifoPath), 0750)).To(Succeed())
			Expect(syscall.Mkfifo(fifoPath, 0600)).To(Succeed())

			resp := doGet("/v1/filesystem?path=/tmp/test.fifo")

			Expect(resp.Code).To(Equal(http.StatusBadRequest))
		})

		It("follows symlink to regular file", func() {
			content := []byte("symlink target content")
			targetPath := filepath.Join(testRootDir, "/tmp/symlink-target.txt")
			linkPath := filepath.Join(testRootDir, "/tmp/symlink.txt")
			Expect(os.MkdirAll(filepath.Dir(targetPath), 0750)).To(Succeed())
			Expect(os.WriteFile(targetPath, content, 0600)).To(Succeed())
			Expect(os.Symlink(targetPath, linkPath)).To(Succeed())

			resp := doGet("/v1/filesystem?path=/tmp/symlink.txt")

			Expect(resp.Code).To(Equal(http.StatusOK))
			Expect(resp.Body.Bytes()).To(Equal(content))
		})

		It("returns 404 for dangling symlink", func() {
			linkPath := filepath.Join(testRootDir, "/tmp/dangling.txt")
			Expect(os.MkdirAll(filepath.Dir(linkPath), 0750)).To(Succeed())
			Expect(os.Symlink("/nonexistent/target", linkPath)).To(Succeed())

			resp := doGet("/v1/filesystem?path=/tmp/dangling.txt")

			Expect(resp.Code).To(Equal(http.StatusNotFound))
		})

		It("returns 422 when path is missing", func() {
			resp := doGet("/v1/filesystem")

			Expect(resp.Code).To(Equal(http.StatusUnprocessableEntity))
		})

		It("returns 422 when path is empty", func() {
			resp := doGet("/v1/filesystem?path=")

			Expect(resp.Code).To(Equal(http.StatusUnprocessableEntity))
		})

		It("returns 500 for null bytes in path", func() {
			resp := doGet("/v1/filesystem?path=/tmp/evil%00file.txt")

			Expect(resp.Code).To(Equal(http.StatusInternalServerError))
		})

		It("succeeds with container specified", func() {
			content := []byte("container read test")
			hostPath := filepath.Join(testRootDir, "/tmp/container-read.txt")
			Expect(os.MkdirAll(filepath.Dir(hostPath), 0750)).To(Succeed())
			Expect(os.WriteFile(hostPath, content, 0600)).To(Succeed())

			resp := doGet("/v1/filesystem?path=/tmp/container-read.txt&container=main")

			Expect(resp.Code).To(Equal(http.StatusOK))
			Expect(resp.Body.Bytes()).To(Equal(content))
		})

		It("normalizes path with dot segments", func() {
			content := []byte("normalized content")
			hostPath := filepath.Join(testRootDir, "/tmp/normalized-read.txt")
			Expect(os.MkdirAll(filepath.Dir(hostPath), 0750)).To(Succeed())
			Expect(os.WriteFile(hostPath, content, 0600)).To(Succeed())

			resp := doGet("/v1/filesystem?path=/tmp/../tmp/./normalized-read.txt")

			Expect(resp.Code).To(Equal(http.StatusOK))
			Expect(resp.Body.Bytes()).To(Equal(content))
		})

		It("reads empty file", func() {
			hostPath := filepath.Join(testRootDir, "/tmp/empty-read.txt")
			Expect(os.MkdirAll(filepath.Dir(hostPath), 0750)).To(Succeed())
			Expect(os.WriteFile(hostPath, []byte{}, 0600)).To(Succeed())

			resp := doGet("/v1/filesystem?path=/tmp/empty-read.txt")

			Expect(resp.Code).To(Equal(http.StatusOK))
			Expect(resp.Body.Len()).To(Equal(0))
		})

		It("sets Content-Type and Content-Length headers", func() {
			content := []byte("header check content")
			hostPath := filepath.Join(testRootDir, "/tmp/headers.txt")
			Expect(os.MkdirAll(filepath.Dir(hostPath), 0750)).To(Succeed())
			Expect(os.WriteFile(hostPath, content, 0600)).To(Succeed())

			resp := doGet("/v1/filesystem?path=/tmp/headers.txt")

			Expect(resp.Code).To(Equal(http.StatusOK))
			Expect(resp.Header().Get("Content-Type")).To(Equal("application/octet-stream"))
		})
	})

	Describe("POST /filesystem", func() {
		It("writes file with absolute path", func() {
			content := []byte("hello world")
			resp := doPost("/v1/filesystem?path=/tmp/test.txt", content)

			Expect(resp.Code).To(Equal(http.StatusNoContent))

			hostPath := filepath.Join(testRootDir, "/tmp/test.txt")
			written, err := os.ReadFile(hostPath) //nolint:gosec // test file path
			Expect(err).NotTo(HaveOccurred())
			Expect(written).To(Equal(content))
		})

		It("writes file with relative path", func() {
			content := []byte("relative file content")
			resp := doPost("/v1/filesystem?path=myfile.txt", content)

			Expect(resp.Code).To(Equal(http.StatusNoContent))

			hostPath := filepath.Join(testRootDir, "/workspace/myfile.txt")
			written, err := os.ReadFile(hostPath) //nolint:gosec // test file path
			Expect(err).NotTo(HaveOccurred())
			Expect(written).To(Equal(content))
		})

		It("creates parent directories", func() {
			content := []byte("nested file")
			resp := doPost("/v1/filesystem?path=/deep/nested/dir/file.txt", content)

			Expect(resp.Code).To(Equal(http.StatusNoContent))

			hostPath := filepath.Join(testRootDir, "/deep/nested/dir/file.txt")
			written, err := os.ReadFile(hostPath) //nolint:gosec // test file path
			Expect(err).NotTo(HaveOccurred())
			Expect(written).To(Equal(content))
		})

		It("returns 422 when path is missing", func() {
			resp := doPost("/v1/filesystem", []byte("some content"))

			Expect(resp.Code).To(Equal(http.StatusUnprocessableEntity))
		})

		It("returns 422 when path is empty", func() {
			resp := doPost("/v1/filesystem?path=", []byte("some content"))

			Expect(resp.Code).To(Equal(http.StatusUnprocessableEntity))
		})

		It("succeeds when container is specified", func() {
			content := []byte("container test")
			resp := doPost("/v1/filesystem?path=/tmp/container-test.txt&container=main", content)

			Expect(resp.Code).To(Equal(http.StatusNoContent))
		})

		It("writes empty file", func() {
			resp := doPost("/v1/filesystem?path=/tmp/empty.txt", []byte{})

			Expect(resp.Code).To(Equal(http.StatusNoContent))

			hostPath := filepath.Join(testRootDir, "/tmp/empty.txt")
			info, err := os.Stat(hostPath)
			Expect(err).NotTo(HaveOccurred())
			Expect(info.Size()).To(Equal(int64(0)))
		})

		It("normalizes path with dot segments", func() {
			content := []byte("normalized")
			resp := doPost("/v1/filesystem?path=/tmp/../tmp/./normalized.txt", content)

			Expect(resp.Code).To(Equal(http.StatusNoContent))
		})

		It("succeeds with empty container name", func() {
			content := []byte("no container specified")
			resp := doPost("/v1/filesystem?path=/tmp/no-container.txt", content)

			Expect(resp.Code).To(Equal(http.StatusNoContent))
		})

		It("returns 500 for null bytes in path", func() {
			content := []byte("malicious content")
			resp := doPost("/v1/filesystem?path=/tmp/evil%00file.txt", content)

			Expect(resp.Code).To(Equal(http.StatusInternalServerError))
		})
	})
})

// errorMockProcFS returns errors for FindMarkedPID.
type errorMockProcFS struct {
	findPIDError error
}

func (m *errorMockProcFS) FindMarkedPID(containerName string) (int, error) {
	return 0, m.findPIDError
}

func (m *errorMockProcFS) GetCwd(pid int) (string, error) {
	return "/workspace", nil
}

func (m *errorMockProcFS) GetRoot(pid int) string {
	return "/tmp"
}

func (m *errorMockProcFS) GetUIDGID(pid int) (int, int, error) {
	return os.Getuid(), os.Getgid(), nil
}

func (m *errorMockProcFS) GetEnviron(pid int) ([]string, error) {
	return []string{"PATH=/usr/bin:/bin", "HOME=/root"}, nil
}

// deadlineCapture is an http.ResponseWriter that records Set*Deadline calls.
// http.ResponseController discovers the methods via interface assertions.
type deadlineCapture struct {
	httptest.ResponseRecorder
	mu             sync.Mutex
	writeDeadlines []time.Time
	readDeadlines  []time.Time
}

func (d *deadlineCapture) SetWriteDeadline(t time.Time) error {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.writeDeadlines = append(d.writeDeadlines, t)
	return nil
}

func (d *deadlineCapture) SetReadDeadline(t time.Time) error {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.readDeadlines = append(d.readDeadlines, t)
	return nil
}

func (d *deadlineCapture) Flush() {
	d.mu.Lock()
	defer d.mu.Unlock()
}

func (d *deadlineCapture) getWriteDeadlines() []time.Time {
	d.mu.Lock()
	defer d.mu.Unlock()
	return append([]time.Time{}, d.writeDeadlines...)
}

func (d *deadlineCapture) getReadDeadlines() []time.Time {
	d.mu.Lock()
	defer d.mu.Unlock()
	return append([]time.Time{}, d.readDeadlines...)
}

// newIsolatedAPI registers the filesystem handlers over a fresh temp root and
// returns the raw handler, for tests that need to supply their own ResponseWriter.
func newIsolatedAPI(prefix string) (http.Handler, string) {
	rootDir, err := os.MkdirTemp("", prefix)
	Expect(err).NotTo(HaveOccurred())
	DeferCleanup(func() { _ = os.RemoveAll(rootDir) })

	Expect(os.MkdirAll(filepath.Join(rootDir, "/workspace"), 0750)).To(Succeed())

	mockProcFS := &MockProcFS{
		rootDir: rootDir,
		cwd:     "/workspace",
		uid:     os.Getuid(),
		gid:     os.Getgid(),
	}

	handler, api := humatest.New(GinkgoT(), huma.DefaultConfig("Test API", "0.1.0"))
	v1 := huma.NewGroup(api, "/v1")
	logger := slog.New(slog.NewTextHandler(GinkgoWriter, nil))
	Register(v1, New(logger, mockProcFS, sandboxsidecar.NewPIDResolver(mockProcFS)))
	return handler, rootDir
}

var _ = Describe("Filesystem deadline wiring", func() {
	var (
		fsHandler http.Handler
		rootDir   string
	)

	BeforeEach(func() {
		fsHandler, rootDir = newIsolatedAPI("deadline-test-*")
	})

	It("sets read and write deadlines during file upload", func() {
		body := make([]byte, 48*1024)
		_, err := rand.Read(body)
		Expect(err).NotTo(HaveOccurred())

		mock := &deadlineCapture{ResponseRecorder: *httptest.NewRecorder()}
		req := httptest.NewRequest("POST", "/v1/filesystem?path=/tmp/deadline-upload.txt", bytes.NewReader(body))
		req.Header.Set("Content-Type", "application/octet-stream")

		fsHandler.ServeHTTP(mock, req)

		Expect(mock.Code).To(Equal(http.StatusNoContent))
		Expect(mock.getReadDeadlines()).NotTo(BeEmpty())
		Expect(mock.getWriteDeadlines()).NotTo(BeEmpty())

		hostPath := filepath.Join(rootDir, "/tmp/deadline-upload.txt")
		written, err := os.ReadFile(hostPath) //nolint:gosec // test file path
		Expect(err).NotTo(HaveOccurred())
		Expect(written).To(Equal(body))
	})

	It("sets write deadlines during file download", func() {
		content := make([]byte, 48*1024)
		_, err := rand.Read(content)
		Expect(err).NotTo(HaveOccurred())

		hostPath := filepath.Join(rootDir, "/tmp/deadline-download.txt")
		Expect(os.MkdirAll(filepath.Dir(hostPath), 0750)).To(Succeed())
		Expect(os.WriteFile(hostPath, content, 0600)).To(Succeed())

		mock := &deadlineCapture{ResponseRecorder: *httptest.NewRecorder()}
		req := httptest.NewRequest("GET", "/v1/filesystem?path=/tmp/deadline-download.txt", nil)

		fsHandler.ServeHTTP(mock, req)

		Expect(mock.Code).To(Equal(http.StatusOK))
		Expect(mock.getWriteDeadlines()).NotTo(BeEmpty())
		Expect(mock.Body.Bytes()).To(Equal(content))
	})
})

// failAfterFirstWrite serves the first write and fails every one after it,
// simulating a stream that dies partway through.
type failAfterFirstWrite struct {
	httptest.ResponseRecorder
	err   error
	wrote bool
}

func (w *failAfterFirstWrite) Write(p []byte) (int, error) {
	if w.wrote {
		return 0, w.err
	}
	w.wrote = true
	return w.ResponseRecorder.Write(p)
}

var _ = Describe("Filesystem stream abort", func() {
	var (
		fsHandler http.Handler
		rootDir   string
	)

	BeforeEach(func() {
		fsHandler, rootDir = newIsolatedAPI("abort-test-*")
	})

	// larger than io.Copy's 32KB buffer so the copy takes more than one write
	writeTestFile := func(name string) {
		content := make([]byte, 64*1024)
		_, err := rand.Read(content)
		Expect(err).NotTo(HaveOccurred())
		hostPath := filepath.Join(rootDir, "/tmp", name)
		Expect(os.MkdirAll(filepath.Dir(hostPath), 0750)).To(Succeed())
		Expect(os.WriteFile(hostPath, content, 0600)).To(Succeed())
	}

	It("aborts the response when a mid-stream write fails", func() {
		writeTestFile("abort.txt")

		mock := &failAfterFirstWrite{ResponseRecorder: *httptest.NewRecorder(), err: errors.New("stream broke")}
		req := httptest.NewRequest("GET", "/v1/filesystem?path=/tmp/abort.txt", nil)

		Expect(func() { fsHandler.ServeHTTP(mock, req) }).To(PanicWith(http.ErrAbortHandler))
		Expect(mock.Body.Len()).To(BeNumerically(">", 0), "expected a partial write before the failure")
	})

	It("ends quietly when the client disconnects mid-stream", func() {
		writeTestFile("disconnect.txt")

		mock := &failAfterFirstWrite{ResponseRecorder: *httptest.NewRecorder(), err: syscall.EPIPE}
		req := httptest.NewRequest("GET", "/v1/filesystem?path=/tmp/disconnect.txt", nil)

		Expect(func() { fsHandler.ServeHTTP(mock, req) }).NotTo(Panic())
	})
})

var _ = Describe("Filesystem error cases", func() {
	var errorAPI humatest.TestAPI

	Describe("container not found", func() {
		BeforeEach(func() {
			logger := slog.New(slog.NewTextHandler(GinkgoWriter, nil))

			mockProcFS := &errorMockProcFS{
				findPIDError: proc.ErrContainerNotFound,
			}

			_, errorAPI = humatest.New(GinkgoT(), huma.DefaultConfig("Error Test API", "0.1.0"))
			v1 := huma.NewGroup(errorAPI, "/v1")
			handler := New(logger, mockProcFS, sandboxsidecar.NewPIDResolver(mockProcFS))
			Register(v1, handler)
		})

		It("returns 400 when container not found", func() {
			resp := errorAPI.Post("/v1/filesystem?path=/tmp/test.txt", "application/octet-stream", bytes.NewReader([]byte("content")))

			Expect(resp.Code).To(Equal(http.StatusBadRequest))
		})
	})
})

File: sdks/typescript/tests/e2e/filesystem.test.ts
// Copyright The Isola Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Mirrors sdks/python/tests/e2e/test_filesystem.py.

import { randomUUID } from "node:crypto";
import { afterAll, beforeAll, describe, expect, it } from "vitest";
import { BadRequestError, Isola, IsolaError, NotFoundError, type Sandbox } from "../../src";
import { ISOLA_URL, safeDelete, waitForRunning } from "./_helpers";

function uniquePath(prefix: string, ext = ".txt"): string {
  return `/tmp/${prefix}_${randomUUID().replaceAll("-", "")}${ext}`;
}

describe.sequential("e2e: filesystem", () => {
  let client: Isola;
  let sandbox: Sandbox;
  const created: string[] = [];

  beforeAll(async () => {
    client = new Isola({ url: ISOLA_URL });
    sandbox = await client.sandboxes.create({ image: "alpine:3.21" });
    created.push(sandbox.id);
    sandbox = await waitForRunning(client, sandbox.id);
  }, 90_000);

  afterAll(async () => {
    for (const id of created) await safeDelete(client, id);
    await client.close();
  });

  it("write+read string round-trips", async () => {
    await sandbox.filesystem.write("/tmp/hello.txt", "hello world\n");
    const bytes = await sandbox.filesystem.read("/tmp/hello.txt");
    expect(new TextDecoder().decode(bytes)).toBe("hello world\n");
  }, 30_000);

  it("write+read bytes round-trips", async () => {
    const payload = new Uint8Array([1, 2, 3, 4, 5, 0, 255, 128]);
    await sandbox.filesystem.write("/tmp/bin.dat", payload);
    const bytes = await sandbox.filesystem.read("/tmp/bin.dat");
    expect(Array.from(bytes)).toEqual(Array.from(payload));
  }, 30_000);

  it("write creates parent directories implicitly", async () => {
    await sandbox.filesystem.write("/tmp/nested/dir/file.txt", "deep");
    const bytes = await sandbox.filesystem.read("/tmp/nested/dir/file.txt");
    expect(new TextDecoder().decode(bytes)).toBe("deep");
  }, 30_000);

  it("write overwrites existing files", async () => {
    await sandbox.filesystem.write("/tmp/over.txt", "first");
    await sandbox.filesystem.write("/tmp/over.txt", "second");
    const bytes = await sandbox.filesystem.read("/tmp/over.txt");
    expect(new TextDecoder().decode(bytes)).toBe("second");
  }, 30_000);

  it("read of missing file raises NotFoundError", async () => {
    await expect(sandbox.filesystem.read("/tmp/does-not-exist.txt")).rejects.toThrow(NotFoundError);
  }, 30_000);

  it("write+read large files", async () => {
    const payload = "x".repeat(200_000);
    await sandbox.filesystem.write("/tmp/big.txt", payload);
    const bytes = await sandbox.filesystem.read("/tmp/big.txt");
    expect(new TextDecoder().decode(bytes).length).toBe(200_000);
  }, 30_000);

  it("commands see files written via filesystem", async () => {
    await sandbox.filesystem.write("/tmp/seen.txt", "via filesystem\n");
    const r = await sandbox.commands.run(["cat", "/tmp/seen.txt"]);
    expect(r.exitCode).toBe(0);
    expect(r.stdout).toBe("via filesystem\n");
  }, 30_000);

  it("filesystem reads files written by commands", async () => {
    await sandbox.commands.run(["sh", "-c", "echo from-command > /tmp/cmd.txt"]);
    const bytes = await sandbox.filesystem.read("/tmp/cmd.txt");
    expect(new TextDecoder().decode(bytes)).toBe("from-command\n");
  }, 30_000);

  // Mirrors test_filesystem.py: write+read binary 256-byte payload.
  it("write+read binary preserves all 256 byte values", async () => {
    const path = uniquePath("write_binary", ".bin");
    const content = new Uint8Array(256 + 5);
    for (let i = 0; i < 256; i++) content[i] = i;
    content.set([0, 0xff, 0x80, 0xfe, 0x01], 256);

    await sandbox.filesystem.write(path, content);
    const result = await sandbox.filesystem.read(path);
    expect(Array.from(result)).toEqual(Array.from(content));
  }, 30_000);

  it("write 1 MiB file round-trips intact", async () => {
    const path = uniquePath("large", ".bin");
    const content = new Uint8Array(1024 * 1024).fill(0x78); // 'x'
    await sandbox.filesystem.write(path, content);
    const result = await sandbox.filesystem.read(path);
    expect(result.length).toBe(content.length);
    // Avoid huge equality check; sample a few bytes.
    expect(result[0]).toBe(0x78);
    expect(result[result.length - 1]).toBe(0x78);
  }, 60_000);

  it("special characters in filename round-trip", async () => {
    const unique = randomUUID().replaceAll("-", "");
    const path = `/tmp/test file (1) ${unique}.txt`;
    const content = "special chars in filename";
    await sandbox.filesystem.write(path, content);
    const result = await sandbox.filesystem.read(path);
    expect(new TextDecoder().decode(result)).toBe(content);
  }, 30_000);

  it("read of a directory returns BadRequest/IsolaError", async () => {
    let caught: unknown;
    try {
      await sandbox.filesystem.read("/tmp");
    } catch (err) {
      caught = err;
    }
    expect(caught).toBeDefined();
    expect(caught instanceof BadRequestError || caught instanceof IsolaError).toBe(true);
  }, 30_000);

  it("uploaded file is executable by commands (cross-subsystem)", async () => {
    const unique = randomUUID().replaceAll("-", "");
    const path = `/tmp/test_${unique}.sh`;
    await sandbox.filesystem.write(path, "#!/bin/sh\necho cross_subsystem_works\n");
    const r = await sandbox.commands.run(["sh", path]);
    expect(r.exitCode).toBe(0);
    expect(r.stdout).toContain("cross_subsystem_works");
  }, 30_000);

  it("relative path is resolved against the container's cwd", async () => {
    const unique = randomUUID().replaceAll("-", "");
    const filename = `relative_${unique}.txt`;
    const content = "relative path test";
    await sandbox.filesystem.write(filename, content);
    const result = await sandbox.filesystem.read(filename);
    expect(new TextDecoder().decode(result)).toBe(content);
  }, 30_000);

  it("writing a file under a path whose parent is a regular file errors", async () => {
    const unique = randomUUID().replaceAll("-", "");
    const blocker = `/tmp/${unique}/blocker`;
    await sandbox.filesystem.write(blocker, "I am a file, not a directory");
    await expect(sandbox.filesystem.write(`${blocker}/nested.txt`, "should fail")).rejects.toThrow(IsolaError);
  }, 30_000);

  it("empty (zero-byte) file write+read", async () => {
    const path = uniquePath("empty");
    await sandbox.filesystem.write(path, "");
    const result = await sandbox.filesystem.read(path);
    expect(result.length).toBe(0);
  }, 30_000);

  it("file ownership matches container UID/GID", async () => {
    // sidecar resolves uid/gid via /proc/<pid>/status, then chown's the file.
    const path = uniquePath("ownership");
    await sandbox.filesystem.write(path, "ownership test");

    const uidR = await sandbox.commands.run(["id", "-u"]);
    const gidR = await sandbox.commands.run(["id", "-g"]);
    const statR = await sandbox.commands.run(["stat", "-c", "%u %g", path]);

    const uid = uidR.stdout.trim();
    const gid = gidR.stdout.trim();
    const [fileUid, fileGid] = statR.stdout.trim().split(" ");
    expect(fileUid).toBe(uid);
    expect(fileGid).toBe(gid);
  }, 30_000);

  it("file written by a command is readable via the filesystem API (nsenter/proc consistency)", async () => {
    const unique = randomUUID().replaceAll("-", "");
    const path = `/tmp/cmd_written_${unique}.txt`;
    const expected = `written_${unique}`;
    const r = await sandbox.commands.run(["sh", "-c", `printf '%s' ${expected} > ${path}`]);
    expect(r.exitCode).toBe(0);
    const content = await sandbox.filesystem.read(path);
    expect(new TextDecoder().decode(content)).toBe(expected);
  }, 30_000);

  it("targeting the primary container by name on filesystem operations", async () => {
    const path = uniquePath("container_param");
    const content = "container param test";
    await sandbox.filesystem.write(path, content, { container: "sandbox0" });
    const result = await sandbox.filesystem.read(path, { container: "sandbox0" });
    expect(new TextDecoder().decode(result)).toBe(content);
  }, 30_000);
});
